### PR TITLE
Improve ClassCondition performance

### LINKED ISSUE #1: Large Html conversion consume much memory
[Original issue](https://github.com/danfickle/openhtmltopdf/issues/962#issue-2096522730) by @Hikariqz

Hi guys,

I'm having an issue where converting a large HTML document to PDF is consuming a significant amount of memory. The HTML represents around 1200 pages, but the file size is only around 3MB.

When generating the PDF, I noticed my Java VM memory usage increasing up to around 1GB. I used a profiler to investigate further and found that a lot of byte arrays were being created during the PDF generation process. Most of the memory usage appeared to be occurring in calls to com.openhtmltopdf.css.newmatch.Condition#matches which involves java.lang.StringBuilder.toString or java.lang.StringBuilder.

image

I'm hoping someone may be able to provide some help or insights on how I could optimize this process to use less memory. Converting such a large number of HTML pages to PDF seems to be straining the memory usage and I want to find a more efficient way to handle it.

Really appreciate it!

### PR DESCRIPTION
**Pull request by jochenberger**
fixes #1

### CODE REVIEW

**Comment by siegelzc** [comment]
When I get the chance, I want to run the profiler with @hikariqz's [test file](https://github.com/openhtmltopdf/openhtmltopdf/files/14076114/test.txt) to compare the two branches.

**Comment by Hikariqz** [comment]
It looks awesome, will test in my locally as well. Thank you all very much @jochenberger @siegelzc !

**Review by siegelzc — CHANGES_REQUESTED** [review]
I tested it with the IntelliJ profiler, and perhaps I'm just using it wrong, but I don't see this memory usage spike. I like the change though, just see my other comment.

File: openhtmltopdf-core/src/main/java/com/openhtmltopdf/css/newmatch/Condition.java [hunk 1]
```diff
@@ -383,20 +383,31 @@ boolean matches(Object e, AttributeResolver attRes, TreeResolver treeRes) {
...
+                return false;
+            }
+            if (idx != 0 && c.charAt(idx - 1) != ' '){
+                return false;
+            }
+            int endOfMatch = idx + _className.length();
+            if (endOfMatch == c.length()){
+                return true;
+            }
+            return endOfMatch < c.length() && c.charAt(endOfMatch) == ' ';
```
[the remark this answers is not in the document]
Comment (jochenberger) [reply openhtmltopdf-core/src/main/java/com/openhtmltopdf/css/newmatch/Condition.java:404]: Go ahead and change it according to your liking then, I don't mind.

**Review by siegelzc — APPROVED** [review]

**Comment by jochenberger** [comment]
I guess the results will depend on your hardware, JVM version, and JVM settings.
Any insights @Hikariqz ?

**Comment by Hikariqz** [comment]
I think so. When I using JDK17 with vm option -Xmx2g -Xms2g, there is clearly memory usage spike. Please find the snippshot 1 below.
image

However, the memory usage is relatively low, almost below 400MB when test it without vm options. Please find the snippshot2 below. (the default max heap size 4GB in my pc).
image

I am guessing the memory spike is cause by jvm option differences, any idea about it?
Will analysis more and get back to you.

**Comment by siegelzc** [comment]
It is interesting to me that there are so many more young generation garbage collection events in the 4gb heap vs the 2gb

**Pull request merged**